core/amber/src/main/scala/edu/uci/ics/amber/engine/common/promise/PromiseManager.scala

```diff
@@ -0,0 +1,221 @@
+
+import scala.collection.mutable
+
+/** This is the central unit of handling promises in actors.
+  * @param selfID
+  * @param controlOutputPort
+  */
+class PromiseManager(selfID: ActorVirtualIdentity, controlOutputPort: ControlOutputPort)
+    extends LazyLogging {
+
+  // the monotonically increasing identifier
+  // for promises created by this promise manager
+  protected var promiseID = 0L
+
+  // promises that created by this promise manager will be put
+  // into this map, send to the destination, and wait for the
+  // return value.
+  protected val unCompletedPromises = new mutable.HashMap[PromiseContext, WorkflowPromise[_]]()
+
+  // this set holds promises for groups.
+  protected val unCompletedGroupPromises = new mutable.HashSet[GroupedWorkflowPromise[_]]()
+
+  // context for current executing promise.
+  protected var promiseContext: PromiseContext = _
+
+  // synchronous promises will be enqueued.
+  protected val queuedInvocations = new mutable.Queue[PromisePayload]()
+
+  // save all synchronous promises with the same root.
+  protected val ongoingSyncPromises = new mutable.HashSet[PromiseContext]()
+
+  // the root context for the current synchronous promise.
+  protected var syncPromiseRoot: PromiseContext = _
+
+  // empty promise handler.
+  // default behavior: discard
+  protected var promiseHandler: PartialFunction[PromiseBody[_], Unit] = {
+    case promise =>
+      logger.info(s"discarding $promise")
+  }
+
+  // process one promise message.
+  def execute(event: PromisePayload): Unit = {
+    event match {
+      // handle return value
+      case ret: ReturnPayload =>
+        // if the return value corresponds to one of the
+        // promises created by this manager.
+        if (unCompletedPromises.contains(ret.context)) {
+          val p = unCompletedPromises(ret.context)
+          // get the context
+          promiseContext = p.ctx
+          ret.returnValue match {
+            case throwable: Throwable =>
+              // if an error is returned
+              p.setException(throwable)
+            case _ =>
+              // otherwise, we put the return value
+              // and invoke the callback.
+              p.setValue(ret.returnValue.asInstanceOf[p.returnType])
+          }
+          // the promise is resolved
+          unCompletedPromises.remove(ret.context)
+        }
+
+        // check if the return value belongs to group promises
+        for (i <- unCompletedGroupPromises) {
+          if (i.takeReturnValue(ret)) {
+            // all return values are collected, set the context
+            promiseContext = i.promise.ctx
+            // invoke callback
+            i.invoke()
+            // the group promise is resolved
+            unCompletedGroupPromises.remove(i)
+          }
+        }
+
+      // handle root synchronous promise
+      case PromiseInvocation(ctx: RootPromiseContext, call: SynchronizedInvocation) =>
```

Comment:
why is call a `SynchronizedInvocation` and not a `PromiseBody`?

Reply:
changed it to `PromiseBody[_] with SynchronizedInvocation`